autobrr: init at 1.57.0

av-gal opened the pull request:
## Description of changes

Reopening #283389. Closes #224560.
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dotlambda commented on maintainers/maintainer-list.nix:
Don't remove others.

av-gal replied:
fixed, sorry!

dotlambda commented on nixos/doc/manual/release-notes/rl-2405.section.md:
No removals please!

dotlambda commented on nixos/modules/module-list.nix:
```diff
@@ -1527,7 +1528,6 @@
   ./tasks/filesystems/jfs.nix
   ./tasks/filesystems/nfs.nix
   ./tasks/filesystems/ntfs.nix
-  ./tasks/filesystems/overlayfs.nix
```
same

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+
+  meta = with lib; {
+    description = "Modern, easy to use download automation for torrents and usenet.";
+    license = licenses.gpl2;
```
`gpl2Only` or `gpl2Plus`?

av-gal replied:
gplv2Plus

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+    license = licenses.gpl2;
+    homepage = "https://autobrr.com/";
+    maintainers = with maintainers; [ av-gal ];
+  };
```
add `meta.mainProgram`

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+  passthru.updateScript = nix-update-script { };
+
+  meta = with lib; {
+    description = "Modern, easy to use download automation for torrents and usenet.";
```
```suggestion
    description = "Modern, easy to use download automation for torrents and usenet";
```

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+  };
+
+  autobrr-web = stdenvNoCC.mkDerivation {
+
```
```suggestion
```

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+  version = "1.36.0";
+  src = fetchFromGitHub {
+    owner = "autobrr";
+    repo = pname;
```
```suggestion
    repo = "autobrr";
```

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+  };
+in
+buildGoModule rec {
+  inherit pname version src autobrr-web;
```
```suggestion
  inherit pname version src;
```
You can put in in `passthru` if you want.

av-gal replied:
I did not :)

dotlambda reviewed (dismissed):
Three commits please:
* maintainers: add av-gal
* autobrr: init at 1.36.0
* nixos/autobrr: init

av-gal committed:
maintainers: add av-gal

av-gal committed:
autobrr: init at 1.57.0

av-gal committed:
nixos/autobrr: init

av-gal reviewed (commented):
done!

av-gal commented:
I think I've addressed all remaining concerns.

dotlambda commented on nixos/modules/services/misc/autobrr.nix:
```diff
@@ -0,0 +1,70 @@
+        DynamicUser = true;
+        StateDirectory = "autobrr";
+        StateDirectoryMode = "0700";
+        ExecStart = "${pkgs.autobrr}/bin/autobrr --config '${WorkingDirectory}'";
```
```suggestion
        ExecStart = "${lib.getExe pkgs.autobrr} --config '${WorkingDirectory}'";
```

dotlambda commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+
+  vendorHash = "sha256-SkwSKFEZAmjVnaSowIbrdH667vB5WqNrPuRs/Yh6BLc=";
+
+  preBuild = "cp -r ${autobrr-web}/* web/dist";
```
```suggestion
  preBuild = ''
    cp -r ${autobrr-web}/* web/dist
  '';
```

ambroisie replied:
For easy diffs, I'd use a multi-line string:

```suggestion
  preBuild = ''
    cp -r ${autobrr-web}/* web/dist
  '';
```

ambroisie commented on nixos/modules/services/misc/autobrr.nix:
```diff
@@ -0,0 +1,70 @@
+        DynamicUser = true;
+        StateDirectory = "autobrr";
+        StateDirectoryMode = "0700";
+        ExecStart = "${pkgs.autobrr}/bin/autobrr --config '${WorkingDirectory}'";
```
Where is `WorkingDirectory` defined?

ambroisie replied:
And please use `lib.getExe`:

```suggestion
        ExecStart = "${lib.getExe pkgs.autobrr} --config '${WorkingDirectory}'";
```

dotlambda reviewed (commented):
There should be a `settings` option to create `config.toml`. See https://github.com/NixOS/rfcs/pull/42.

ambroisie commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,71 @@
+    hash = "sha256-HQuiREj84nTv9EpKaFKNmCLLmHRko3QR4qoa6ICWwHY=";
+  };
+
+  autobrr-web = stdenvNoCC.mkDerivation {
+    pname = "${pname}-web";
+    inherit src version;
+
+    nativeBuildInputs = [
+      nodePackages.pnpm
+      cacert
+    ];
+
+    buildPhase = ''
+      runHook preBuild
+
+      export HOME=$(mktemp -d)
+      pnpm --dir web install
+      pnpm --dir web run build
+
+      mv web/dist $out
+
+      runHook postBuild
+    '';
+
+    dontInstall = true;
+
+    outputHashMode = "recursive";
+    outputHash = "sha256-IOrW26Nq+9PYWlzUSpPBfkv3jzs5VlfF0JVFaUCDMmw=";
+  };
```
So at least we're not using an FOD to store the `pnpm` dependencies. But I still worry that the output is not bit-for-bit reproducible.

My recommendation is to:
1. Try to use `pnpm-export` and use `buildNpmPackage` or `mkYarnPackage`.
2. If `pnpm-export` doesn't work, just generate a lock file manually.

In either case, this would mean vendoring the lock file to the repository.

av-gal replied:
I believe this derivation should be bit-for-bit reproducible. Different versions of pnpm might have different store or cache formats, but each module in `node_modules` will be consistent with its hash in `pnpm-lock.yaml`. `dist` just contains the final bundle created by Vite. Why would that change because the dependencies are grabbed with a different package manager?

dotlambda replied:
What makes you believe the output of `pnpm --dir web run build` is reproducible? Do TypeScript and Vite claim to be reproducible?

dotlambda replied:
```
error: build of '/nix/store/cfcl0gpb9p5nnphw25g9mc6hifjpf3wl-autobrr-web-1.36.0.drv' on 'ssh://nix@willard' failed: hash mismatch in fixed-output derivation '/nix/store/cfcl0gpb9p5nnphw25g9mc6hifjpf3wl-autobrr-web-1.36.0.drv':
         specified: sha256-IOrW26Nq+9PYWlzUSpPBfkv3jzs5VlfF0JVFaUCDMmw=
            got:    sha256-SkwSKFEZAmjVnaSowIbrdH667vB5WqNrPuRs/Yh6BLc=
```

av-gal replied:
`pnpm --dir web run build` just runs `tsc && vite build`, as specified in `package.json`. Neither of these claim to be reproducible build tools (I don't think), but I believe that `buildNpmPackage` or `mkYarnPackage` would execute those commands in the same way. Please correct me if I'm wrong.

av-gal replied:
May I ask if you did anything to cause the hash mismatch? Or did you just try to build it on your machine

dotlambda replied:
The latter.



> Neither of these claim to be reproducible build tools (I don't think), but I believe that `buildNpmPackage` or `mkYarnPackage` would execute those commands in the same way.

Yes but those aren't used in FODs.

nu-nu-ko commented on nixos/modules/services/misc/autobrr.nix:
```diff
@@ -0,0 +1,70 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
```
consider `inherits` where you find yourself using `lib.*` excessively.
[this](https://nix.dev/guides/best-practices.html#with-scopes)  nix.dev page on best practices with `with` may be helpful
an issue (#208242) has been made to track inappropriate uses of `with` such as this.

nu-nu-ko commented on nixos/modules/services/misc/autobrr.nix:
```diff
@@ -0,0 +1,70 @@
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc "Open ports in the firewall for the Autobrr web interface.";
```
```suggestion
        description = "Open ports in the firewall for the Autobrr web interface.";
```
`lib.mdDoc` is now just an alias and can be safely removed everywhere.
see https://github.com/NixOS/nixpkgs/commit/d36f950d405369abca5144b14a201bc25c4a080a and https://github.com/NixOS/nixpkgs/pull/237557

zze0s commented:
We just released v1.38.0 so might be a good idea to bump the version in this pr :smile:

SuperSandro2000 commented on nixos/modules/services/misc/autobrr.nix:
```diff
@@ -0,0 +1,70 @@
+  config = mkIf cfg.enable {
+    systemd.services.autobrr = {
+      description = "Autobrr";
+      after = [ "syslog.target" "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
```
```suggestion
      after = [ "syslog.target" "network-online.target" ];
      wants = [ "network-online.target" ];
      wantedBy = [ "multi-user.target" ];
```

muradbu commented:
Is this still being worked on?

av-gal commented:
I plan on picking this up again shortly.

zze0s commented:
Great! The move to pnpm was the most recent big build related changes but other than that it should be smooth sailing.

Current version is 1.50.0 :)

muradbu commented:
> I plan on picking this up again shortly.

Can I assist you with anything?

av-gal commented:
I think I've got it mostly figured out. I'm just not sure how to handle the session secret. Ideally autobrr's config.toml would be created from the NixOS module's settings, but I believe this would cause it to be added to the nix store. Which is problematic, since the nix store is world-readable, and that file also contains the secret for cookie authentication.

BatteredBunny commented:
> Any suggestions on how to avoid this?

https://github.com/autobrr/autobrr/blob/e2df1e4fb8e501fbf1b81b1878c2ae472bf7d62b/internal/config/config.go#L246-L354
I think optimal is to add an `environmentFile` option and specify the secrets in there, since autobrr has env var support

BatteredBunny commented on nixos/modules/services/misc/autobrr.nix:
```diff
@@ -0,0 +1,76 @@
+        ExecPreStart =
+          "${lib.getExe pkgs.dasel} put -f '${configTemplate}' -v '${
+            builtins.readFile cfg.secretFile
+          }' -o %S/config.toml 'sessionSecret'";
```
FYI ``serviceConfig.EnvironmentFile = cfg.secretFile;`` might work better

BatteredBunny replied:
Also if nix is reading the file doesn't it go into nix store? (Didn't test somebody correct me if not)

av-gal replied:
You might be right about builtins.readFile adding files to the Nix store (I am also not sure), but I have used shell substitution instead. I don't want to use the environment variables because they must be retyped by users in the `EnvironmentFile`, and systemd doesn't have a way to prevent users from specifying settings in the environment which conflict with those in the module's configuration.

av-gal replied:
I also cannot find any documentation for autobrr's environment variables outside of the code, so I don't want to point users towards them.

BatteredBunny replied:
https://autobrr.com/installation/docker#environment-variables
They are mentioned here in the documentation. Also from what i have seen, env files for secrets are a common pattern in nixpkgs

BatteredBunny replied:
Anyway, apart from my opinion on this, looks good now 👍

av-gal commented:
I think this is ready for review.

gepbird reviewed (changes requested):
Thanks for your contribution, it is very much appreciated!

Apart from the nits (which you are welcome to ignore if you think they are not that important), the remaining things to do is fixing the update script and the nixos test.

av-gal commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,83 @@
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = {
+    description = "Modern, easy to use download automation for torrents and usenet";
+    license = lib.licenses.gpl2Plus;
+    homepage = "https://autobrr.com/";
+    maintainers = with lib.maintainers; [ av-gal ];
+    mainProgram = "autobrr";
+  };
```
I'm not sure what to put here. Do I match what is built upstream, or set `lib.platforms.all` in the hopes that it will work on more exotic OSs as well?

av-gal replied:
I went with what gets built upstream: Darwin. FreeBSD, Linux, and Windows.

gepbird replied:
I'm not sure whether only the tested platforms or the platforms supported by upstream should go here, it's inconsistent in many other packages too. This looks fine :)

av-gal commented:
Addressed most of your suggestions @gepbird, thanks for the feedback. Still need to fix the update script, and a new issue  related to fetchGithub not retrieving the commit hash when it downloads the repository archive. Marking as this as draft again until I get those issues fixed.

gepbird commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,93 @@
+    changelog = "https://autobrr.com/release-notes/v${version}";
+    maintainers = with lib.maintainers; [ av-gal ];
+    mainProgram = "autobrr";
+    platforms = with lib.platforms; [
+      darwin
+      freebsd
+      linux
+      windows
+    ];
```
Trying to build the package fails this way:
<details><summary>eval log</summary>

```console
error: Package ‘autobrr-1.56.1’ in /nix/store/4vpb2ky00pl6l7kz5wmhnh0bcpms786a-source/pkgs/by-name/au/autobrr/package.nix:80 is not available on the requested hostPlatform:
  hostPlatform.config = "x86_64-unknown-linux-gnu"
  package.meta.platforms = [
    [
      "x86_64-darwin"
      "i686-darwin"
      "aarch64-darwin"
      "armv7a-darwin"
    ]
    [
      "i686-freebsd"
      "x86_64-freebsd"
      "aarch64-freebsd"
    ]
    [
      "aarch64-linux"
      "armv5tel-linux"
      "armv6l-linux"
      "armv7a-linux"
      "armv7l-linux"
      "i686-linux"
      "loongarch64-linux"
      "m68k-linux"
      "microblaze-linux"
      "microblazeel-linux"
      "mips-linux"
      "mips64-linux"
      "mips64el-linux"
      "mipsel-linux"
      "powerpc64-linux"
      "powerpc64le-linux"
      "riscv32-linux"
      "riscv64-linux"
      "s390-linux"
      "s390x-linux"
      "x86_64-linux"
    ]
    [
      "i686-cygwin"
      "x86_64-cygwin"
      "aarch64-windows"
      "x86_64-windows"
      "i686-windows"
    ]
  ]
```
</details>

Because for example `lib.platforms.linux` is already a list (`[ "x86_64-linux" "aarch64-linux" ... ]`), you need to concatinate these lists

```suggestion
    platforms = with lib.platforms; darwin ++ freebsd ++ linux ++ windows;
```

gepbird commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,93 @@
+    "-X main.version=${version}"
+    "-X main.commit=${src.tag}"
+  ];
+
+  passthru.updateScript = nix-update-script {
```
An extra thing you could add is checking the version of one of the binaries:

```suggestion
  doInstallCheck = true;
  nativeInstallCheckInputs = [
    versionCheckHook
  ];
  versionCheckProgram = "${placeholder "out"}/bin/autobrrctl";
  versionCheckProgramArg = "version";

  passthru.updateScript = nix-update-script {
```

av-gal replied:
`autobrrctl version` checks for the most recent version and prints it after the current version. So I think `versionCheckHook` would always succeed, even if the actual version is incorrect.

gepbird replied:
> `autobrrctl version` checks for the most recent version and prints it after the current version. So I think `versionCheckHook` would always succeed, even if the actual version is incorrect.

In a regular build that's correct, but thankfully the nix sandbox disallows networking during most of the build time. From the output you can see the local version and it failing to fetch the version:

```console
autobrr> Executing versionCheckPhase                                                                                                                                                                              
autobrr> Successfully managed to find version 1.56.1 in the output of the command /nix/store/nmvccjllygw5l539gmcdra6il0nzb34c-autobrr-1.56.1/bin/autobrrctl version                                               
autobrr> Version: 1.56.1                                                                                                                                                                                          
autobrr> Commit: v1.56.1                                                                                                                                                                                          
autobrr> Build:                                                                                                                                                                                                   
autobrr> Failed to fetch latest release from api: Get "https://api.autobrr.com/repos/autobrr/autobrr/releases/latest": dial tcp: lookup api.autobrr.com on [::1]:53: read udp [::1]:58973->[::1]:53: read: connect
ion refused                                                                                                                                                                                                       
autobrr> Finished versionCheckPhase  
```

gepbird reviewed (commented):
Other than the above issue, changes LGTM!

Tested the update script, building the package, running the test, basic functionality of the binaries.

zze0s commented:
Yesterday we released [v1.57.0](https://github.com/autobrr/autobrr/releases/tag/v1.57.0) No changes to the build etc.

gepbird reviewed (approved):
Tested as before, I think this is ready :)

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 287593`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>autobrr</li>
  </ul>
</details>

av-gal commented:
@SuperSandro2000 friendly ping :)

FliegendeWurst commented on pkgs/by-name/au/autobrr/package.nix:
```diff
@@ -0,0 +1,96 @@
+    changelog = "https://autobrr.com/release-notes/v${version}";
+    maintainers = with lib.maintainers; [ av-gal ];
+    mainProgram = "autobrr";
+    platforms = with lib.platforms; darwin ++ freebsd ++ linux ++ windows;
```
```suggestion
    platforms = with lib.platforms; darwin ++ freebsd ++ linux;
```
Does not cross-compile to Windows.
`certstore/certstore_windows.go:640:29: cannot convert cmsg (variable of type *_Ctype_char) to type _Ctype_HLOCAL`
